ruma-api-macros/src/api/request/outgoing.rs

```diff
@@ -131,23 +131,29 @@ impl Request {
             .collect();
 
         for auth in &metadata.authentication {
-            if auth.value == "AccessToken" {
-                let attrs = &auth.attrs;
-                header_kvs.extend(quote! {
+            let attrs = &auth.attrs;
+            let if_required_expr = if auth.value == "AccessToken" {
+                quote! { Some(access_token) }
+            } else {
+                quote! { None }
+            };
+            header_kvs.extend(quote! {{
+                let access_token = match access_token {
+                    #ruma_api::SendAccessToken::IfRequired(access_token) => #if_required_expr,
+                    #ruma_api::SendAccessToken::Always(access_token) => Some(access_token),
+                    #ruma_api::SendAccessToken::No => None,
+                };
+
+                if let Some(access_token) = access_token {
                     #( #attrs )*
                     req_headers.insert(
                         #http::header::AUTHORIZATION,
                         #http::header::HeaderValue::from_str(
-                            &::std::format!(
-                                "Bearer {}",
-                                access_token.ok_or(
-                                    #ruma_api::error::IntoHttpError::NeedsAuthentication
```

Comment:
Looking at the overall logic again, it seems like we have lost the logic that if the user doesn't specify an access token (now `SendAccessToken::No`) but one is required according to the endpoint definition, we return `Err(NeedsAuthentication)` without even sending the request.

Reply:
My interpretation, and what the docs for `SendAccessToken::No` seem to suggest, is that `SendAccessToken::No` is supposed to simply never add an `access_token` even if the endpoint requires it. While `SendAccessToken::IfRequired` adds the given `access_token` only if the endpoint requires it, and none otherwise. So it seems there is no `IntoHttpError::NeedsAuthentication` any longer?

Reply:
Maybe `SendAccessToken::Never` would be more clear?

Reply:
> My interpretation, and what the docs for `SendAccessToken::No` currently suggest, is that `SendAccessToken::No` is supposed to simply never add an `access_token` even if the endpoint requires it.

I'm not sure we have a use case for that though. What `access_token: None` currently means is that the client simply doesn't have an access token yet, i.e. the user is not logged in. Maybe we need to have a different enum name than `SendAccessToken` to make this clearer, but that's the only thing I've been able to come up with.

Reply:
So `SendAccessToken::None` with documentation and code saying, that if the endpoint requires the token, it'll return an error?

Reply:
If you think that's better than `SendAccessToken::No`, sure.

Reply:
Updated. Yeah, I feel like `None` better suggests that an access_token simply isn't available (in combination with the enum name), while `No` sounds more like, no, you never going to send that access token :D